src/mod_vcard_rdbms.erl

```diff
@@ -69,130 +92,271 @@ get_vcard(LUser, LServer) ->
-    SMiddle = mongoose_rdbms:escape_string(VCardSearch#vcard_search.middle),
-    SLMiddle = mongoose_rdbms:escape_string(VCardSearch#vcard_search.lmiddle),
-    SNickname = mongoose_rdbms:escape_string(VCardSearch#vcard_search.nickname),
-    SLNickname = mongoose_rdbms:escape_string(VCardSearch#vcard_search.lnickname),
-    SBDay = mongoose_rdbms:escape_string(VCardSearch#vcard_search.bday),
-    SLBDay = mongoose_rdbms:escape_string(VCardSearch#vcard_search.lbday),
-    SCTRY = mongoose_rdbms:escape_string(VCardSearch#vcard_search.ctry),
-    SLCTRY = mongoose_rdbms:escape_string(VCardSearch#vcard_search.lctry),
-    SLocality = mongoose_rdbms:escape_string(VCardSearch#vcard_search.locality),
-    SLLocality = mongoose_rdbms:escape_string(VCardSearch#vcard_search.llocality),
-    SEMail = mongoose_rdbms:escape_string(VCardSearch#vcard_search.email),
-    SLEMail = mongoose_rdbms:escape_string(VCardSearch#vcard_search.lemail),
-    SOrgName = mongoose_rdbms:escape_string(VCardSearch#vcard_search.orgname),
-    SLOrgName = mongoose_rdbms:escape_string(VCardSearch#vcard_search.lorgname),
-    SOrgUnit = mongoose_rdbms:escape_string(VCardSearch#vcard_search.orgunit),
-    SLOrgUnit = mongoose_rdbms:escape_string(VCardSearch#vcard_search.lorgunit),
-
-    rdbms_queries:set_vcard(VHost,
-                           SLServer, SLUsername, SBDay, SCTRY, SEMail,
-                           SFN, SFamily, SGiven, SLBDay, SLCTRY,
-                           SLEMail, SLFN, SLFamily, SLGiven,
-                           SLLocality, SLMiddle, SLNickname,
-                           SLOrgName, SLOrgUnit, SLocality,
-                           SMiddle, SNickname, SOrgName,
-                           SOrgUnit, SVCARD, SUsername),
-
-    mongoose_hooks:vcard_set(VHost, ok, LUser, VCard),
-    ok.
+    SearchArgs = assert_binaries(search_args(User, Search)),
+    XML = exml:to_binary(VCard),
+    F = fun() ->
+            update_vcard_t(LUser, LServer, XML),
+            update_vcard_search_t(LUser, LServer, SearchArgs),
+            ok
+        end,
+    Result = handle_result(rdbms_queries:sql_transaction(LServer, F)),
+    log_upsert_result(LServer, LUser, VCard, XML, Result),
+    Result.
 
-search(LServer, Data) ->
-    RestrictionSQL = make_restriction_sql(LServer, Data),
-    R = do_search(LServer, RestrictionSQL),
-    lists:map(fun(I) -> record_to_item(LServer, I) end, R).
-
-do_search(_LServer, "") ->
-    [];
-do_search(LServer, RestrictionSQL) ->
-    Limit = mod_vcard:get_results_limit(LServer),
-    case catch rdbms_queries:search_vcard(LServer, RestrictionSQL, Limit) of
-        {selected, Rs} when is_list(Rs) ->
-            Rs;
-        Error ->
-            ?LOG_ERROR(#{what => vcard_db_search_failed,
-                         reason => Error, host => LServer}),
-            []
+%% Horrible things happen to unicode list of bytes in MySQL driver, when
+%% making prepared queries.
```

Comment:
Like what things? 🤔

Reply:
Corrupted data would be inserted into the database.

Reply:
Can you explain why? Also, "Horrible things happen to" is a pretty useless and confusing comment. You can actually explain in the comment that "If XXX then corrupted data would be inserted because of YYY"

Reply:
```erlang
1> X = <<"Москва"/utf8>>.
<<208,156,208,190,209,129,208,186,208,178,208,176>>
2> U = unicode:characters_to_list(X).
[1052,1086,1089,1082,1074,1072]
3> unicode:characters_to_binary(X).
<<208,156,208,190,209,129,208,186,208,178,208,176>>
4> unicode:characters_to_binary(U).
<<208,156,208,190,209,129,208,186,208,178,208,176>>
5> unicode:characters_to_binary(binary_to_list(X)). % data corruption here because of invalid conversation
<<195,144,194,156,195,144,194,190,195,145,194,129,195,144,
  194,186,195,144,194,178,195,144,194,176>>
```

Reply:
Yes, new comment is much better 👍